### DISCUSSION TITLE
Vulnerability in quarkus-core dependency

### DISCUSSION THREAD

**Discussion by expertesantos** [quarkusio/quarkus#43607] [Q&A] [answered]
Hi everyone,

I’m facing an issue with Quarkus. The company I work for is strict about vulnerabilities, and AWS Amazon Inspector has pointed out that the Quarkus project is using the library org.jboss.logging.jboss-logging-3.6.0.Final.jar, which relies on the log4j version 1.2.17, leading to several vulnerabilities being flagged: CVE-2019-17571, CVE-2022-23305, CVE-2023-26464, CVE-2022-23302, CVE-2022-23307, CVE-2021-4104, CVE-2020-9488.

I tried to exclude the jboss-logging dependency, but this library is used in quarkus-core, and I encountered the following error:

```
org/jboss/logging/BasicLogger: java.lang.NoClassDefFoundError
java.lang.NoClassDefFoundError: org/jboss/logging/BasicLogger
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(Unknown Source)
        at java.base/java.security.SecureClassLoader.defineClass(Unknown Source)
        at java.base/java.net.URLClassLoader.defineClass(Unknown Source)
        at java.base/java.net.URLClassLoader$1.run(Unknown Source)
        at java.base/java.net.URLClassLoader$1.run(Unknown Source)
        at java.base/java.security.AccessController.doPrivileged(Unknown Source)
        at java.base/java.net.URLClassLoader.findClass(Unknown Source)
        at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
        at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
        at io.smallrye.config.SmallRyeConfig$ConfigSources.getSources(SmallRyeConfig.java:968)
        at io.smallrye.config.SmallRyeConfig$ConfigSources.<init>(SmallRyeConfig.java:804)
        at io.smallrye.config.SmallRyeConfig.<init>(SmallRyeConfig.java:86)
        at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:736)
        at io.quarkus.runtime.generated.Config.<clinit>(Unknown Source)
        at io.quarkus.runner.ApplicationImpl.<clinit>(Unknown Source)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Unknown Source)
        at io.quarkus.runtime.Quarkus.manualInitialize(Quarkus.java:236)
        at io.quarkus.amazon.lambda.runtime.QuarkusStreamHandler.<init>(QuarkusStreamHandler.java:21)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Unknown Source)
        at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
Caused by: java.lang.ClassNotFoundException: org.jboss.logging.BasicLogger
        at java.base/java.net.URLClassLoader.findClass(Unknown Source)
        at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
        at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
        ... 25 more
```

I saw that there’s already a discussion on the jboss-logging project, but there hasn’t been any movement: https://github.com/jboss-logging/jboss-logging/discussions/111

Considering that I cannot deploy with AWS Amazon Inspector raising this issue, do you have any suggestions?

**Comment by phillip-kruger** [comment]
//cc @dmlloyd

**Reply by dmlloyd** [reply]
It does not in fact actually include `log4j`. It is only in the dependency list with a `provided` scope, meaning that `jboss-logging` may be used with systems which use `log4j` for logging. Quarkus is not such a system.

**Comment by dmlloyd** [comment, marked as answer]
Since we do not include `log4j`, this is not an issue (note that the dependency is `provided` scope). Unfortunately these vulnerability scanners don't always understand the nature of a dependency. But, `jboss-logging` cannot be said to depend on `log4j` in any logical sense.

**Comment by nicklasweasel** [comment]
Hi, did you file a ticket with AWS or find any other workarounds? I turned on Lambda scanning and now I also have multiple pages of "critical" findings...

**Reply by expertesantos** [reply]
Hi, the response I received was that they do not agree that it's a false positive because even though it's marked as 'provided', it is still a potential risk.

Unfortunately, we are having to rethink the use of Quarkus as we follow strict security policies.

**Reply by expertesantos** [reply]
Just to add, I agree that it is a false positive, as the log4j jar is not in my lib folder and is also not listed in the output of the mvn dependency:tree command.

But as you know, it's uncomfortable to have so many alerts since they appear for each lambda.

**Reply by nicklasweasel** [reply]
It pulls down your security score pretty hard since they're ranked critical

**Reply by gsmet** [reply]
I can understand your point but you are using tools that are not exactly doing their job well (`provided` dependencies shouldn't be taken into consideration as they are not included) so us having the burden of working around their limitations is not exactly attractive. And what is even "better" from our perspective is that each tool has its own set of limitations.

Now in this case, it might be possible to update the Log4j version in there but I will defer to @dmlloyd and @jamezp as to checking if it's a good idea or not.
Given nobody should use this particular version in the wild, I think it makes sense to update to a version that is reasonably secure - even if not stricly required from a security perspective.

**Reply by dmlloyd** [reply]
We could possibly make it be `optional` as well (since, well, it is).

**Reply by dmlloyd** [reply]
I want to note that there are zero non-vulnerable versions of log4j 1.x, so there is no update we can do to mitigate that.  We could possibly consider dropping support for log4j 1.x at some point.

**Reply by expertesantos** [reply]
If you make the change, I can test it in the vulnerability scan. Regarding the Optional option, we can test it, but I believe it would be the same case as the Provided scope.

**Reply by jamezp** [reply]
FWIW WildFly has dropped log4j 1.x support. There is https://reload4j.qos.ch/ and I can say log4j-jboss-logmanager did remove the vulnerable parts of log4j 1.x.

**Reply by dmlloyd** [reply]
Hey @jamezp, if `log4j-jboss-logmanager` has enough of the API, could we possibly use that (provided+optional) instead of log4j proper as `jboss-logging`'s dependency on that API?

**Reply by dmlloyd** [reply]
Good idea, `reload4j` is another option we could pursue.

**Reply by gsmet** [reply]
+1 for reload4j given it's apparently binary compatible.

**Reply by jamezp** [reply]
> Hey @jamezp, if `log4j-jboss-logmanager` has enough of the API, could we possibly use that (provided+optional) instead of log4j proper as `jboss-logging`'s dependency on that API?

You should be able to assuming the jboss-logmanager is available too :) The only thing I don't know is if security scanners also flag it. I'm not sure what they look for. The types [are not shaded in](https://github.com/jboss-logging/log4j-jboss-logmanager/blob/09a08ed8c059dcf75b57d741b6c3809282ada84f/pom.xml#L149-L153), so hopefully it wouldn't trigger any scanners vulnerabilities.

**Reply by dmlloyd** [reply]
Oh, maybe a simpler solution would be a `*`/`*` exclusion in the `dependencyManagement` of the build parent?

**Reply by dmlloyd** [reply]
I've drafted this in #43861 but it's not clear whether it's actually a good idea (specifically, whether it's a better idea than having better dependencies in `jboss-logging` or making them optional or something else).

**Reply by gsmet** [reply]
> Oh, maybe a simpler solution would be a */* exclusion in the dependencyManagement of the build parent?

@dmlloyd do you mean on the Quarkus side?

@expertesantos is your report per dependency or it's a global report for your app and if our BOM exclude the dependencies entirely, that might work? Maybe you could experiment a bit with it and report back?

**Reply by dmlloyd** [reply]
Yes on the Quarkus side. Which is not ideal (because we repeat separate dep mgmt for that library many times) but it might be a good enough bandage.

**Reply by expertesantos** [reply]
> > Oh, maybe a simpler solution would be a _/_ exclusion in the dependencyManagement of the build parent?
> 
> @dmlloyd do you mean on the Quarkus side?
> 
> @expertesantos is your report per dependency or it's a global report for your app and if our BOM exclude the dependencies entirely, that might work? Maybe you could experiment a bit with it and report back?

@gsmet It's a scan that analyzes the dependencies of my application, which is a Lambda (serverless).

I'm not sure if I understood, but I did the following test:

- I cloned the repository https://github.com/quarkusio/quarkus;
- I made the changes from the PR https://github.com/quarkusio/quarkus/pull/43861;
- I ran the command mvn clean install -DskipTests in the root directory of the project;
- In my project, I changed the Quarkus version to 999-SNAPSHOT;
- I deployed my project and ran the scan;
- The Log4j vulnerability is still appearing.
- A new vulnerability has been listed: CVE-2011-4605 - org.jboss.naming in the library lib/org.jboss.jboss-transaction-spi-8.0.0.Final.jar/META-INF/maven/org.jboss/jboss-transaction-spi/pom.xml.

**Reply by dmlloyd** [reply]
Ah. Unfortunately then the only conclusion I can arrive at is that your scanner is fatally broken. It's reporting artifacts as risks which are not actually in the project. There is no logical basis for such an error. This is similar to flagging an error because the string `"log4j"` appears somewhere in the program text - or because there are an even number of bytes in an artifact. It does not indicate anything of use.

It is generally not possible to accommodate illogical software. We can throw things at the wall all day but nothing we do will be "correct", and we would risk breaking other things in the meantime even if we did hit on the magical combination that allowed the scan to pass. I think an upstream bug fix on the scanner is the only way forward.

In the meantime, I would recommend filtering out any dependency which is not present in the final artifact.

**Reply by expertesantos** [reply]
Regarding changing the use of log4j to reload4j in the jboss-logging project, is that not an option?

**Reply by nicklasweasel** [reply]
The absurd thing is that if you deploy the lambda as native, the "problem" goes away. Hell, you could even have the vulnerable log4j compiled in there and the scanner wouldn't detect it.

However, I agree with @expertesantos that it's a real problem for some organizations and for people trying to get their system pass the AWS FTR, projects that Quarkus would probably want to have as references.

**Reply by dmlloyd** [reply]
> Regarding changing the use of log4j to reload4j in the jboss-logging project, is that not an option?

Sure, we can definitely try it and see what happens. Overall though I think the responsibility for accuracy lies with the maintainers of the scanner.
